### PR TITLE
Initial implementation of deploy command

### PR DESCRIPTION
**Pull request by tpdownes**
Support workflows for deploying single- or multi-group blueprints via the following sequence of commands:

```text
ghpc create blueprint.yaml
ghpc deploy deployment_root_directory
```

No direct user execution of Terraform or Packer is necessary. If the optional `--auto-approve` flag is supplied to `deploy`, then changes to cloud infrastructure are automatically applied. If not supplied, the user must approve every deployment group separately.

In this initial implementation, `ghpc` blindly iterates over all deployment groups in-order. It will only attempt to run "terraform apply" if "terraform plan -detailed-exitcode" returns code 2. It will always attempt to build Packer images. For each group, it will always attempt to invoke "import-inputs" and "export-outputs". These are operationally harmless on groups without inputs or exports, but may print log entries to stdout.

Example output:

```text
❯❯❯ ghpc create -w --vars project_id=project examples/hpc-cluster-small.yaml
Terraform group 'primary' was successfully created in directory hpc-small/primary
To deploy, run the following commands:

terraform -chdir=hpc-small/primary init
terraform -chdir=hpc-small/primary validate
terraform -chdir=hpc-small/primary apply

❯❯❯ ghpc deploy hpc-small
2023/05/10 14:31:05 initializing terraform directory hpc-small/primary
2023/05/10 14:31:23 testing if terraform state of hpc-small/primary is in sync with cloud infrastructure
2023/05/10 14:31:30 cloud infrastructure requires changes
2023/05/10 14:31:30 Summary of proposed changes: Plan: 13 to add, 0 to change, 0 to destroy.
Display full proposed changes, Apply proposed changes, Stop and exit, Continue without applying? [d,a,s,c]: a
2023/05/10 14:31:35 running terraform apply on group hpc-small/primary
module.slurm_login.module.slurm_cluster_login_node.data.google_compute_default_service_account.default: Reading...
##
## MANY LINES OF TERRAFORM APPLY OUTPUT
##
Apply complete! Resources: 13 added, 0 changed, 0 destroyed.
2023/05/10 14:36:22 collecting terraform outputs from hpc-small/primary
2023/05/10 14:36:23 group primary contains no artifacts to export
❯❯❯ ghpc deploy hpc-small --auto-approve
2023/05/10 14:45:04 testing if terraform state of hpc-small/primary is in sync with cloud infrastructure
2023/05/10 14:45:08 cloud infrastructure requires no changes
2023/05/10 14:45:08 collecting terraform outputs from hpc-small/primary
2023/05/10 14:45:08 group primary contains no artifacts to export
```

### Submission Checklist

Please take the following actions before submitting this pull request.

* Fork your PR branch from the Toolkit "develop" branch (not main)
* Test all changes with pre-commit in a local branch [#](https://goo.gle/hpc-toolkit#development)
* Confirm that "make tests" passes all tests
* Add or modify unit tests to cover code changes
* Ensure that unit test coverage remains above 80%
* Update all applicable documentation
* Follow Cloud HPC Toolkit Contribution guidelines [#](https://goo.gle/hpc-toolkit-contributing)

### CODE REVIEW

File: pkg/shell/terraform.go
Comment (nick-stroud) [inline pkg/shell/terraform.go:158]: I think `no` should stop the process. If we want a skip function as was discussed that should be a third option.

Currently if I say not it continue to try to deploy the next groups:
```
2023/05/10 23:58:17 initializing terraform directory igc-01/one
2023/05/10 23:58:22 testing if terraform state of igc-01/one is in sync with cloud infrastructure
2023/05/10 23:58:24 cloud infrastructure requires changes
Do you want to deploy group igc-01/one [y/n]: n
2023/05/10 23:58:28 collecting terraform outputs from igc-01/one
2023/05/10 23:58:28 group one contains no artifacts to export
2023/05/10 23:58:28 collecting outputs for group two from group one
Error: consider running "ghpc export-outputs igc-01/one" (detailed error below)
The configuration file "igc-01/.ghpc/artifacts/one_outputs.tfvars" could not be read.
```

I think it would be reasonable to spit out the same exit as from export when changes are needed:

```
2023/05/11 00:01:12 testing if terraform state of igc-01/one is in sync with cloud infrastructure
2023/05/11 00:01:14 cloud infrastructure requires changes
Error: cloud infrastructure requires changes; please run "terraform -chdir=igc-01/one apply"
```

Comment (nick-stroud) [inline pkg/shell/terraform.go:158]: I feel there would be a lot of value to print the plan before asking if we should apply. If there is a way for us to add this, even if it is not efficient to start, I think we should as it totally changes the user experience.

My recommendation to not overburden this initial implementation would be to call plan again, now to output, under the circumstances that changes are needed and we need user approval. 

I think you could simply add the following before asking the user to approve:

```
		log.Printf("generating plan for group %s", tf.WorkingDir())
		tf.SetStdout(os.Stdout)
		tf.SetStderr(os.Stderr)
		tf.Plan(context.Background()) // guessing on syntax
		tf.SetStdout(nil)
		tf.SetStderr(nil)
```

User output would look like:

```
initializing terraform directory igc-01/one
testing if terraform state of igc-01/one is in sync with cloud infrastructure
cloud infrastructure requires changes
generating plan for group igc-01/one
...
... many lines of terraform plan
...
Do you want to deploy group igc-01/one [y/n]:
```

Then we can optimize so we are not running plan twice in future pr.

File: pkg/shell/terraform.go
Comment (nick-stroud) [inline pkg/shell/terraform.go:153]: nit: Remove timestamp from output lines. If you do not do this in this PR, please open a task.

File: cmd/deploy.go
Comment (nick-stroud) [inline cmd/deploy.go:109]: nit: Probably for future PR. We should print out destroy commands (and sequence) at end of deploy. We are now doing so much for the user, I suspect it will be hard for them to figure out what to do to destroy. Open task to close this.

File: cmd/deploy.go
Comment (nick-stroud) [inline cmd/deploy.go:1]: nit: for `^C` behavior, I think the right thing to do would be to at least try to pass kill to terraform. Not needed for this PR but please open a task.

File: cmd/deploy.go
Comment (tpdownes) [reply cmd/deploy.go:39]: I think it is likely that future iterations of this command are going to need some significant custom validation logic no matter what. This is exposed through "PreRun/PreRunE". There is already an example of a simple PreRun hook here.

e.g. I think it's easy to say "these flags are mutually exclusive" but my quick read does not suggest an easy way of saying "but you must supply 1".

So, from a technical perspective of validating the inputs, I think all options are about the same amount of work. Different choices (esp. your suggestion) would probably require us to write custom usage/help functions.

File: cmd/export.go
Comment (tpdownes) [reply cmd/export.go:91]: It's now the responsibility of the caller to validate the existence of a group that has been specified actually exists. Some commands target a group (e.g. export-outputs, import-inputs, future versions of this command) while others (e.g. this one) target a deployment directory and then iterate over groups. So this `func` was doing more than is generally useful and I found it better to have the blueprint available to the callers anyways.

I anticipate extending `shell.ValidateDeploymentDirectory` to verify kinds (terraform/packer) based on its best guess of the kind after scanning the directory.

File: pkg/shell/terraform.go
Comment (tpdownes) [reply pkg/shell/terraform.go:158]: I will alter this to have a 3d skip-and-continue option.

Comment (tpdownes) [reply pkg/shell/terraform.go:158]: I can do this, but will probably choose the "optimal" approach of directing the plan to a file and running apply on the file. Lots of good reasons and not obviously much more effort than running twice.

File: cmd/deploy.go
Comment (tpdownes) [reply cmd/deploy.go:109]: I agree with the problem and think the answer is to create a destroy command. Packer images an obvious edge case.

Comment (tpdownes) [reply cmd/deploy.go:109]: Creating task.

File: cmd/deploy.go
Comment (tpdownes) [reply cmd/deploy.go:35]: Resolved offline to avoid introducing a short option. Reasoning:

1. Explicitness / alignment with terraform.
2. Easier to add in future than to take away.

File: pkg/config/config.go
Comment (mr0re1) [inline pkg/config/config.go:156]: `s/groupName/n/` it's too verbose and "ungoolangy"

File: pkg/shell/terraform.go
Comment (tpdownes) [reply pkg/shell/terraform.go:158]: I've chosen the following approach:

- capture the stdout of plan (in human-readable format)
- prompt the user with 4 options: `Display proposed changes, Apply proposed changes, Stop and exit, Continue without applying? [d,a,s,c]`
- Loop forever until user picks 1 of the 4 options.

Logic behind making display of the plan optional is that it is very long and causes me to become disoriented if printed immediately to screen.

There is a sub-optimality in that there does not appear to be an existing func in tfexec that will capture both human and machine readable format for the plan. So there is no guarantee that the apply stage will attempt _exactly_ the changes being proposed. I will address this with a task for future work.

File: pkg/config/config.go
Comment (tpdownes) [reply pkg/config/config.go]: Addressed here and in `GroupIndex` since the type now conveys the argument sufficiently.

File: cmd/deploy.go
Comment (tpdownes) [reply cmd/deploy.go]: Resolved in favor of not requiring a flag. Future likely to look like:

```
ghpc deploy [--group group_name] deployment_directory_or_blueprint_yaml
```

Where group_name specifies the group up to which you are willing to deploy.

File: pkg/shell/terraform.go
Comment (tpdownes) [reply pkg/shell/terraform.go]: Continuing this conversation at https://github.com/GoogleCloudPlatform/hpc-toolkit/pull/1304#discussion_r1191575429 and resolving this conversation.

Comment (nick-stroud) [reply pkg/shell/terraform.go]: I disagree with requiring the user to take action before they are told what is about to change (in this case the `d` option). In addition we are now providing the user with 4 options here.

Terraform's native behavior is to always print the delta. I don't believe we should deviate from this behavior without good reason. 

I could see at some point in the future only printing the summary line (`x added, x modified, x deleted`) and having the user ask for the full output, but I don't think that should be our starting point.

**Review by nick-stroud — CHANGES_REQUESTED** [review]

File: pkg/shell/terraform.go
Comment (nick-stroud) [reply pkg/shell/terraform.go]: @cboneti could you help weight in here as a tiebreaker.

Comment (cboneti) [reply pkg/shell/terraform.go]: both of you  are bringing important points here. 
I believe it would be important to give a sense of the changes, so this is what I propose, do one of the two options as a starting point. Pick what you prefer or whatever is easier: 
1) Print, as Nick pointed, a summary line as described below: 
```
❯❯❯ ghpc deploy hpc-small
2023/05/10 14:31:05 initializing terraform directory hpc-small/primary
2023/05/10 14:31:23 testing if terraform state of hpc-small/primary is in sync with cloud infrastructure
2023/05/10 14:31:30 cloud infrastructure requires changes:

Plan: 13 to add, 0 to change, 0 to destroy.

Display proposed changes, Apply proposed changes, Stop and exit, Continue without applying? [d,a,s,c]:
...
```

or, if it is easier this way: 
2) always print terraform plan output and then adjust the prompt accordingly (notice changes after tf plan): 
```
❯❯❯ ghpc deploy hpc-small
2023/05/10 14:31:05 initializing terraform directory hpc-small/primary
2023/05/10 14:31:23 testing if terraform state of hpc-small/primary is in sync with cloud infrastructure
##
## MANY LINES OF TERRAFORM PLAN OUTPUT
##
2023/05/10 14:31:30 cloud infrastructure requires changes for hpc-small/primary
Apply proposed changes, Stop and exit, Continue without applying? [a,s,c]: a
2023/05/10 14:31:35 running terraform apply on group hpc-small/primary
module.slurm_login.module.slurm_cluster_login_node.data.google_compute_default_service_account.default: Reading...
...
```

Comment (tpdownes) [reply pkg/shell/terraform.go]: Offline discussion resulted in choosing (1). 3c598f29 includes:

- printing a summary of changes immediately upon prompt
- continue to allow an optional full summary by entering `d`

Because it was simple within this effort of work, I added the functionality of applying the plan actually being shown to the user rather than "crossing your fingers" that nothing has changed by the time apply is run.

Edge cases addressed:

- Packer full and summary of changes are the same
- Future task will be created to use `PlanJSON` and parse the output for the [change_summary](https://developer.hashicorp.com/terraform/internals/machine-readable-ui#change-summary) field instead of regexp for `Plan:`. If no match is found, an advisory to print the full proposed changes is written to screen.

Comment (cboneti) [reply pkg/shell/terraform.go]: LGTM, please assign to Nick

Comment (nick-stroud) [reply pkg/shell/terraform.go]: LGTM. Just want to test then will approve.

**Review by nick-stroud — APPROVED** [review]

**Pull request merged**
